common/src/ipc/serde.rs

```diff
@@ -0,0 +1,631 @@
+use std::mem::size_of;
+use std::mem::size_of_val;
+use std::num::NonZeroI32;
+use std::num::NonZeroU8;
+use std::time::Duration;
+
+use super::types2::Animation;
+use super::types2::BitPack;
+use super::types2::ClearRequest;
+use super::types2::Coord;
+use super::types2::Image;
+use super::types2::ImageDescription;
+use super::types2::ImageRequest;
+use super::types2::Info;
+use super::types2::PixelFormat;
+use super::types2::Scale;
+use super::types2::Transition;
+use super::types2::TransitionType;
+use super::types2::Vec2;
+
+/// Type for managing access in [`Serialize`] and [`Deserialize`]
+pub struct Cursor<T> {
+    buf: T,
+    pos: usize,
+}
+
+impl<T> Cursor<T> {
+    /// Create [`Self`] from provided buffer
+    pub fn new(buf: T) -> Self {
+        Self { buf, pos: 0 }
+    }
+
+    /// Extract buffer back, discarding all internal state.
+    pub fn finish(self) -> T {
+        self.buf
+    }
+}
+
+impl<'a> Cursor<&'a mut [u8]> {
+    fn write(&mut self, bytes: &[u8]) {
+        let next = self.pos + bytes.len();
+        self.buf[self.pos..next].copy_from_slice(bytes);
+        self.pos = next;
+    }
+
+    fn write_tagged(&mut self, bytes: &[u8]) {
+        bytes.tag().serialize(self);
+        self.write(bytes);
+    }
+}
+
+impl<'a> Cursor<&'a [u8]> {
+    fn read(&mut self, count: usize) -> &'a [u8] {
+        let read = &self.buf[self.pos..self.pos + count];
+        self.pos += count;
+        read
+    }
+
+    fn read_tagged(&mut self) -> &'a [u8] {
+        let count = u32::deserialize(self);
+        self.read(count as usize)
+    }
+}
+
+trait Tagged {
+    fn tag(&self) -> u32;
+}
+
+impl<T> Tagged for [T] {
+    fn tag(&self) -> u32 {
+        self.len() as u32
+    }
+}
+
+impl Tagged for str {
+    fn tag(&self) -> u32 {
+        self.as_bytes().tag()
+    }
+}
+
+/// Serializes data structure into byte slice.
+pub trait Serialize {
```

Comment:
I wouldn't be against it, @LGFae expressed that he wouldn't want to include it. See [comment](https://github.com/LGFae/swww/pull/331#issuecomment-2184159415).

Reply:
Huh. @LGFae could we revisit that? Serde does, by default, copy a lot of data around, but utilizing slices and `#[serde(borrow)]` should allow you to avoid all of that. Or is it something else that's an issue?

Reply:
I think it would simplify the codebase a great deal. +1 from me.

Reply:
As June wrote, there is a [crate](https://lib.rs/crates/postcard), which has a similar format to my current one. Usually, you can just derive desired traits, so no major refactoring needed.

Reply:
Ah, my bad, I thought `postcard` was like, a standard format like `protobuf` or something. It seems to be a lot more flexible than that.

Sure, we can try it out! But I do think we should do it in a separate PR, because I feel like it will be a pretty big change.